lib/compilation-queue.ts

```diff
@@ -41,33 +41,59 @@ const queueCompleted = new PromClient.Counter({
     name: 'ce_compilation_queue_completed_total',
     help: 'Total number of jobs completed',
 });
+const queueStale = new PromClient.Counter({
+    name: 'ce_compilation_queue_stale_total',
+    help: 'Total number of jobs abandoned before starting as they were stale',
+});
 
 export type Job<TaskResultType> = () => PromiseLike<TaskResultType>;
 
+export type EnqueueOptions = {
+    abandonIfStale?: boolean;
+};
+
 export class CompilationQueue {
     private readonly _running: Set<number> = new Set();
     private readonly _queue: Queue;
+    private readonly _staleAfterMs: number;
 
-    constructor(concurrency: number, timeout: number) {
+    constructor(concurrency: number, timeout: number, staleAfterMs: number) {
         this._queue = new Queue({
             concurrency,
             timeout,
             throwOnTimeout: true,
         });
+        this._staleAfterMs = staleAfterMs;
     }
 
     static fromProps(ceProps) {
-        return new CompilationQueue(ceProps('maxConcurrentCompiles', 1), ceProps('compilationEnvTimeoutMs'));
+        return new CompilationQueue(
+            ceProps('maxConcurrentCompiles', 1),
+            ceProps('compilationEnvTimeoutMs'),
+            ceProps('compilationStaleAfterMs', 60_000),
+        );
     }
 
-    enqueue<Result>(job: Job<Result>): PromiseLike<Result> {
+    enqueue<Result>(job: Job<Result>, options?: EnqueueOptions): PromiseLike<Result> {
         const enqueueAsyncId = executionAsyncId();
+        const enqueuedAt = Date.now();
+
         // If we're asked to enqueue a job when we're already in a async queued job context, just run it.
         // This prevents a deadlock.
         if (this._running.has(enqueueAsyncId)) return job();
         queueEnqueued.inc();
         return this._queue.add(() => {
+            const dequeuedAt = Date.now();
             queueDequeued.inc();
+            if (options && options.abandonIfStale && dequeuedAt > enqueuedAt + this._staleAfterMs) {
+                queueCompleted.inc();
+                queueStale.inc();
+                const queueTimeSecs = (dequeuedAt - enqueuedAt) / 1000;
+                const limitSecs = this._staleAfterMs / 1000;
+                throw new Error(
```

Comment:
how does the queue handler handle exceptions, console log and continue?

Reply:
Yes.